src/ucp/proto/proto_select.c

```diff
@@ -0,0 +1,336 @@
+static ucs_status_t
+ucp_proto_select_init_protocols(ucp_worker_h worker,
+                                ucp_worker_cfg_index_t ep_cfg_index,
+                                ucp_worker_cfg_index_t rkey_cfg_index,
+                                const ucp_proto_select_param_t *select_param,
+                                ucp_proto_select_init_protocols_t *proto_init)
+{
+    ucp_proto_init_params_t init_params;
+    ucs_string_buffer_t strb;
+    size_t priv_size, offset;
+    ucp_proto_id_t proto_id;
+    ucs_status_t status;
+    void *tmp;
+
+    ucs_assert(ep_cfg_index != UCP_WORKER_CFG_INDEX_NULL);
+
+    init_params.worker        = worker;
+    init_params.select_param  = select_param;
+    init_params.ep_config_key = &worker->ep_config[ep_cfg_index].key;
+
+    if (rkey_cfg_index == UCP_WORKER_CFG_INDEX_NULL) {
+        init_params.rkey_config_key = NULL;
+    } else {
+        init_params.rkey_config_key = &worker->rkey_config[rkey_cfg_index].key;
+
+        /* rkey configuration must be for the same ep */
+        ucs_assertv_always(
+                init_params.rkey_config_key->ep_cfg_index == ep_cfg_index,
+                "rkey->ep_cfg_index=%d ep_cfg_index=%d",
+                init_params.rkey_config_key->ep_cfg_index, ep_cfg_index);
+    }
+
+    proto_init->select_param = select_param;
+    proto_init->mask         = 0;
+
+    /* Initialize protocols and get their capabilities */
+    proto_init->priv_buf = ucs_malloc(ucp_protocols_count * UCP_PROTO_PRIV_MAX,
+                                      "ucp_proto_priv");
+    if (proto_init->priv_buf == NULL) {
+        status = UCS_ERR_NO_MEMORY;
+        goto err;
+    }
+
+    offset = 0;
+    for (proto_id = 0; proto_id < ucp_protocols_count; ++proto_id) {
+        init_params.priv      = UCS_PTR_BYTE_OFFSET(proto_init->priv_buf,
+                                                          offset);
+        init_params.priv_size  = &priv_size;
+        init_params.caps       = &proto_init->caps[proto_id];
+        init_params.proto_name = ucp_proto_id_field(proto_id, name);
+
+        status = ucp_proto_id_call(proto_id, init, &init_params);
+        if (status != UCS_OK) {
+            continue;
+        }
+
+        proto_init->mask                  |= UCS_BIT(proto_id);
+        proto_init->priv_offsets[proto_id] = offset;
+        offset                            += priv_size;
+    }
+
+    if (proto_init->mask == 0) {
+        /* No protocol can support the given selection parameters */
+        ucp_proto_select_param_str(select_param, &strb);
+        ucs_debug("no protocols found for %s", ucs_string_buffer_cstr(&strb));
+        ucs_string_buffer_cleanup(&strb);
+        status = UCS_ERR_NO_ELEM;
+        goto err_free_priv;
+    }
+
+    /* Finalize the shared priv buffer size */
+    if (offset == 0) {
+        ucs_free(proto_init->priv_buf);
+        proto_init->priv_buf = NULL;
+    } else {
+        tmp = ucs_realloc(proto_init->priv_buf, offset, "ucp_proto_priv");
```

Comment:
does protocol keep an address to priv space (as it was passed to init function by `ucp_proto_id_call`)? This address can be changed by realloc

Reply:
(discussed offline) the pointer is assigned at later stage which did not fit into this PR